Value set: lift offset from numeric constants to expressions

tautschnig opened the pull request:
We can safely track arbitrary expressions as pointer offsets rather than limit ourselves to just constant offsets (and then treating all other expressions as "unknown").

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Value set: lift offset from numeric constants to expressions

We can safely track arbitrary expressions as pointer offsets rather than
limit ourselves to just constant offsets (and then treating all other
expressions as "unknown").

tautschnig committed:
Ensure test only fails as originally designed

The test will otherwise exhibit undefined behaviour under CBMC 6
settings (where malloc may fail). This, in turn, can cause the expected
patterns not to show up in the trace.

remi-delmas-3000 commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -123,20 +121,7 @@ renamedt<ssa_exprt, L2> goto_symex_statet::assignment(
     propagation.erase_if_exists(l1_identifier);
 
-  {
-    // update value sets
-    exprt l1_rhs(rhs);
-    get_l1_name(l1_rhs);
-
-    const ssa_exprt l1_lhs = remove_level_2(lhs);
-    if(run_validation_checks)
-    {
-      DATA_INVARIANT(!check_renaming_l1(l1_lhs), "lhs renaming failed on l1");
-      DATA_INVARIANT(!check_renaming_l1(l1_rhs), "rhs renaming failed on l1");
-    }
-
```
Why is it safe to remove these checks ?

tautschnig replied:
We do not rename back to L1 (what already was renamed to L2), so there is nothing left to check.

remi-delmas-3000 commented on src/goto-symex/shadow_memory_util.cpp:
```diff
@@ -981,7 +981,7 @@ normalize(const object_descriptor_exprt &expr, const namespacet &ns)
     return expr;
   }
-  if(expr.offset().id() == ID_unknown)
+  if(!expr.offset().is_constant())
```
Can we get a high level description of what is the normal form we're trying to reach ?

peterschrammel replied:
To root object + constant offset, as object can be an arbitrary access path into the object. Pointer equality checks become trivial then - maybe simplify_expr has become good enough in the meanwhile.
Anyways, this seems orthogonal to this PR.

remi-delmas-3000 commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -184,7 +183,7 @@ void value_sett::output(std::ostream &out, const std::string &indent) const
 
         if(o_it->second)
-          stream << *o_it->second;
+          stream << format(*o_it->second);
```
now we have to print an expression instead of a mere integer

tautschnig replied:
Yes, but why is that a concern?

remi-delmas-3000 reviewed (approved):
I have one remaining question: Now that we have symbolic offsets for pointer expressions instead of just "constants" or "unknown", is there ay way to use that to compute  more precise results in `get_value_set_rec` ? I know it lets us be more precise when modelling assignments, but I don't understand why we don't have a similar gain in precision when computing dereferences/traversing value_sets. 

Other question: now that the value set representation "knows" that an expression `array[i]` has an offset of the form `i * sizeof(T)` could we try to take into account extra constraints about `i` _during_ the value set traversal ? Let's say we're trying to resolve `array[i]` in the context of a basic loop invariant `0 <= i && i <= len(array)`,  knowing range constraints about `i` we could maybe avoid injecting values representing OOB accesses in the value set for `array[i]` ?

peterschrammel commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -362,7 +361,8 @@ bool value_sett::eval_pointer_offset(
           return false;
 
-        *ptr_offset += *it->second;
+        *ptr_offset +=
```
Codecov says that 361-368 are not covered by any tests.

tautschnig replied:
Done.

peterschrammel commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -717,24 +722,20 @@ void value_sett::get_value_set_rec(
-            if(!i.has_value())
-              i = mp_integer{0};
-            i = *i + *offset;
+            additional_offset = plus_exprt{
```
Codecov says that 731-733 are not covered by any tests.

tautschnig replied:
Done.

peterschrammel commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -745,18 +746,22 @@ void value_sett::get_value_set_rec(
+          additional_offset = mult_exprt{
+            *additional_offset, from_integer(*size, additional_offset->type())};
 
           if(expr.id()==ID_minus)
```
Codecov says that 758-766 are not covered by any tests.

tautschnig replied:
Done.

peterschrammel commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -1393,22 +1410,24 @@ void value_sett::get_reference_set_rec(
             o.reset();
           else
-            *o = *i * (*size);
+          {
```
Codecov says that 1416-1431 are not covered by any tests.

tautschnig replied:
Done.

peterschrammel reviewed (approved):
It seems we are lacking tests in 4 places here. Given that this is all but trivial it would be great to find some test cases that trigger these.

tautschnig commented:
> I have one remaining question: Now that we have symbolic offsets for pointer expressions instead of just "constants" or "unknown", is there ay way to use that to compute more precise results in `get_value_set_rec` ? I know it lets us be more precise when modelling assignments, but I don't understand why we don't have a similar gain in precision when computing dereferences/traversing value_sets.
> 

I have already seen this happen, although it isn't necessarily very obvious unless one starts examining the formula that symex produces. #8653 is a consequence of my observations: I was surprised to still find "unknown" when I had expected a known offset

> Other question: now that the value set representation "knows" that an expression `array[i]` has an offset of the form `i * sizeof(T)` could we try to take into account extra constraints about `i` _during_ the value set traversal ? Let's say we're trying to resolve `array[i]` in the context of a basic loop invariant `0 <= i && i <= len(array)`, knowing range constraints about `i` we could maybe avoid injecting values representing OOB accesses in the value set for `array[i]` ?

I'm not sure we even do create those OOB values here?

tautschnig committed:
value_sett: replace uses of `throw` by invariants

None of these should be triggered by user-provided code.

tautschnig committed:
Additional unit tests for value_sett

Several branches cannot easily be tested by regression tests as
existing front-ends will never create some of the expressions.
